vpr/src/place/place_delay_model.h

```diff
@@ -4,50 +4,95 @@
 #include "vtr_ndmatrix.h"
 #include "vtr_flat_map.h"
 #include "vpr_types.h"
-#include "router_lookahead_map.h"
+#include "router_delay_profiling.h"
 
 //Abstract interface to a placement delay model
 class PlaceDelayModel {
   public:
     virtual ~PlaceDelayModel() = default;
 
+    // Computes place delay model.
+    virtual void compute(
+        const RouterDelayProfile& router,
+        const t_placer_opts& placer_opts,
+        const t_router_opts& router_opts,
+        int longest_length)
+        = 0;
+
     //Returns the delay estimate between the specified block pins
+    //
+    // Either compute or read methods must be invoked before invoking
+    // delay.
     virtual float delay(int from_x, int from_y, int from_pin, int to_x, int to_y, int to_pin) const = 0;
 
     //Dumps the delay model to an echo file
     virtual void dump_echo(std::string filename) const = 0;
+
+    // Write place delay model to specified file.
+    // May be unimplemented, in which case method should throw an exception.
+    virtual void write(const std::string& file) const = 0;
+
+    // Read place delay model to specified file.
+    // May be unimplemented, in which case method should throw an exception.
+    virtual void read(const std::string& file) = 0;
 };
 
 //A simple delay model based on the distance (delta) between block locations
 class DeltaDelayModel : public PlaceDelayModel {
   public:
+    DeltaDelayModel(){};
     DeltaDelayModel(vtr::Matrix<float> delta_delays, t_router_opts router_opts)
         : delays_(std::move(delta_delays))
         , router_opts_(router_opts) {}
 
+    void compute(
+        const RouterDelayProfile& router,
+        const t_placer_opts& placer_opts,
+        const t_router_opts& router_opts,
+        int longest_length) override;
     float delay(int from_x, int from_y, int /*from_pin*/, int to_x, int to_y, int /*to_pin*/) const override;
     void dump_echo(std::string filepath) const override;
 
+    void read(const std::string& file) override {
+        (void)file;
+        VPR_THROW(VPR_ERROR_ROUTE, "DeltaDelayModel::read unimplemented");
+    }
+    void write(const std::string& file) const override {
+        (void)file;
+        VPR_THROW(VPR_ERROR_ROUTE, "DeltaDelayModel::write unimplemented");
+    }
+
   private:
     vtr::Matrix<float> delays_;
     t_router_opts router_opts_;
 };
 
 class OverrideDelayModel : public PlaceDelayModel {
   public:
-    OverrideDelayModel(std::unique_ptr<PlaceDelayModel> base_delay_model, t_router_opts router_opts)
-        : base_delay_model_(std::move(base_delay_model))
-        , router_opts_(router_opts) {}
-
+    void compute(
+        const RouterDelayProfile& router,
```

Comment:
For the various places this is used I think that calling it the argument/variable `router` is a bit misleading (it doesn't return paths, only delays; or put another way it uses the router but is not the router itself). Probably better to call it something like `route_profiler`.

Reply:
Sure, fixed